lib/ansible/utils/vars.py

```diff
@@ -85,11 +85,15 @@ def combine_vars(a, b, merge=None):
 
     if merge or merge is None and C.DEFAULT_HASH_BEHAVIOUR == "merge":
         return merge_hash(a, b)
-    else:
-        # HASH_BEHAVIOUR == 'replace'
-        _validate_mutable_mappings(a, b)
-        result = a | b
-        return result
+
+    # HASH_BEHAVIOUR == 'replace'
+    _validate_mutable_mappings(a, b)
+    # NOTE: In the case of ANSIBLE_DEBUG=1 task_vars is VarsWithSources(MutableMapping)
+    # so '|' operator cannot be used as it can be used only on dicts
+    # https://peps.python.org/pep-0584/#what-about-mapping-and-mutablemapping
+    result = a.copy()
```

Comment:
I would modify `VarsWithSources` to act more like a dict instead (iirc `__or__` )

Reply:
I was thinking the same thing.  Would need both `__or__` and `__ior__`

Reply:
Done. Take a look.